sql, roachpb: correctly marshall retryable errors through DistSQL

andreimatei opened the pull request:
cc @nvanbenschoten

andreimatei committed:
sql, roachpb: correctly marshall retryable errors through DistSQL

Retryable errors are recognized by both the TxnCoordSender and by the
sql.Executor, which react to them. When these errors flow through
DistSQL, the TxnCoordSender is bypassed and the Executor, before this
patch, was seeing an opaque error it wouldn't recognize.

This patch adds proper marshalling for these structured errors in the
form of a new proto and provides a way for the distSQLReceiver to update
the client.Txn's proto on errors.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15187)
<!-- Reviewable:end -->

andreimatei commented:
sending for review while I figure out how to test this sucker

bdarnell commented:
Reviewed 1 of 2 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/dist_sender_server_test.go, line 1152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiArNqDPjWsEQ23ATkl:-KiArNqDPjWsEQ23ATkm:bja6p0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ca10fa19b59a6cb56df3fab904e9e9d446f96b/pkg/kv/dist_sender_server_test.go#L1152)):*
> ```Go
> 				if _, ok := retErr.Cause.(*roachpb.ReadWithinUncertaintyIntervalError); ok {
> 					if !retErr.Transaction.Writing {
> 						t.Errorf("unexpected non-writing txn on error")
> ```

Why did you remove this?

---

*[pkg/roachpb/errors.proto, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiAqrvNC9w5O_AoMu5x:-KiAqrvNC9w5O_AoMu5y:buvueh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4720e3523b5469da22865e58a3f832da04ce0ce5/pkg/roachpb/errors.proto#L98)):*
> ```protobuf
> // In contrast with other errors, the Transaction that the client gets in the
> // pErr carrying this ErrorDetail is not supposed to be used by the client,
> // besides checking that the transaction's ID is righ.
> ```

s/righ/right/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
distsql part LGTM

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KiBPYEQW-RQRdsXkOG-:b23qjo)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
what I've done here is wholly insufficient; I'm ignoring the other updates performed by the `TxnCoordSender` to the `Transaction`, besides calling `Transaction.UpdateStateOnErr`. Stay tuned.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/errors.proto, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiAqrvNC9w5O_AoMu5x:-KiB3nQAkb8hrtHKYnNM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4720e3523b5469da22865e58a3f832da04ce0ce5/pkg/roachpb/errors.proto#L98)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/righ/right/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KiBetj3VAejOLRrazxJ:b-rwnswo)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
client, kv: correctly restart the Transaction proto on DistSQL retryable errors

- TxnCoordSender becomes the only place that deals with restarting a
Transaction. It takes the information from the error described above and
sometimes calls Transaction.Restart(). The role of the client.Txn is
reduced, and restarting-like things are consolidated in the
TxnCoordSender.

In the non-DistSQL world, the flow of a retryable txn becomes:

- storage returns a retryable error such as TransactionRetryError or
TransactionRestartedError. These are ErrorDetails in a pErr, and
pErr.TransactionRestart != NONE identifies them.
- TxnCoordSender.Send() recognizes these errors, takes the error's
Transaction in consideration too, performs Transaction.Restart() and
prepares a restarted Transaction proto.
- TxnCoordSender then puts the restarted Transaction in a
HandledRetryableError. This error is an ErrorDetail, and it's used for
the TxnCoordSender to communicate with the client.Txn.Send() through the
Sender interface. The restarted txn is also copied to its map of ongoing
transactions. This way, the map is kept in sync with the Transaction
that the client.Txn will use.
- client.Txn now only deals with HandledRetryableError's, in terms of
retryable errors. How it handles them is straight-forward: it copies the
Transaction from them to its copy of the Proto (this is the proto that
will be used for all future requests.

On the DistSQL path, the original pErr was transformed by a
pErr.GoError(), but that GoError() return value retains enough
information about the cause such that the restart can be properly done
on the gateway. This error is marshalled to the gateway, which uses a
backdoor into client.Txn to perform the restart and update the
client.Txn state.

andreimatei commented:
Pushed another commit that makes distSQL compensate for the TxnCoordSender. PTAL

r3 is #15305 

---

Review status: 0 of 22 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kv/dist_sender_server_test.go, line 1152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiArNqDPjWsEQ23ATkl:-KiXWaARbi9LZrXEPQTI:bi1wze5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ca10fa19b59a6cb56df3fab904e9e9d446f96b/pkg/kv/dist_sender_server_test.go#L1152)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why did you remove this?
</blockquote></details>

what I had uploaded was incomplete here. The type of the error changes and code like this is no longer supposed to peek inside the error. But the test still tests this `Writing` bit where it matters.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KiXXQYWlnLLW6u57xZn:bqr99nu)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
A test with a DistSQL restart is still forthcoming.

---

Review status: 0 of 22 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KiXXw6Nd030cxPV6QJd:b9bn6o6)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
This makes me nervous. I've always been skeptical about the push to combine client.Txn and kv.TxnCoordSender because client.Txn was fairly thin and didn't have that much overlap with TxnCoordSender. This change introduces significant overlap between them (so I guess it demonstrates the necessity of the merge), and now distsql writes updates to client.Txn that only very indirectly make it in to TxnCoordSender (TxnCoordSender won't update its txn proto until the next command *returns*, I think). 

I think I'd be more comfortable if we could make `client.Txn` reach into `DB.Sender` to call `TxnCoordSender.updateState` instead of introducing more duplicative logic in `client.Txn`

---

Reviewed 4 of 12 files at r4, 18 of 18 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/txn.go, line 78 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaEkj21IghojcoPDVS:-KiaEkj21IghojcoPDVT:b-qlzxqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/internal/client/txn.go#L78)):*
> ```Go
> 		commandCount int
> 
> 		// onKVTxnAborted, if set, is called when a KV-level roachpb.Transaction has
> ```

s/Aborted/Switched/

---

*[pkg/internal/client/txn.go, line 936 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaMA4K4l6ETtzgJ3Nh:-KiaMA4K4l6ETtzgJ3Ni:br0b4a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/internal/client/txn.go#L936)):*
> ```Go
> // higher priority.
> //
> // TODO(andrei,nvanbenschoten): document whether calling this repeatedly within
> ```

It seems like this needs to be resolved before this can be merged. I think we need to make sure that the result of PrepapreTransactionForRetry moves the txn forward in time before assigning it to txn.mu.Proto: it either changes the ID or increases the epoch. 

---

*[pkg/internal/client/txn_test.go, line 588 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaIQftYORnn4MM-JZ1:-KiaIQftYORnn4MM-JZ2:bxj53c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/internal/client/txn_test.go#L588)):*
> ```Go
> 						if count == 1 {
> 							var pErr *roachpb.Error
> 							if _, ok := test.err.(*roachpb.ReadWithinUncertaintyIntervalError); ok {
> ```

Why did this need to be added here? Have we lost something on the non-test path?

---

*[pkg/roachpb/errors.proto, line 350 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaJPE5AYmTaJJ3ngqX:-KiaJPE5AYmTaJJ3ngqY:bxpqiro) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/roachpb/errors.proto#L350)):*
> ```protobuf
>   // will also not be set.
>   //
>   // NOTE(andrei): I think a better design would be for the server to always be
> ```

Could we move towards this design today by having this error contain a `Transaction` and a `bool aborted` instead of either a `Transaction` or a set of `txn_id, retry_priority, retry_timestamp`?  If `aborted` is true, only those fields would be filled in (and `PrepareTransactionForRetry` would know this)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KiaS0GSSkhlpplIebEQ:b-4wawdl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
https://github.com/cockroachdb/cockroach/pull/15254

is depending on this PR to provide a mechanism to freshly retry a transaction within txn.Exec() when the closure run by Exec() returns a retryable error from the sql layer (getTableLease() can push the transaction forward). We need code that will create a new txn with a new id, and which will abort the previous incarnation of the transaction. Ideally we would like the intents to also be GC-ed but that's not a hard requirement given that we don't expect this retry to happen very often (only on txns running concurrent with a schema change potentially)

andreimatei commented:
Ben, I've moved the restarting logic back into TxnCoordSender. PTAL.

---

Review status: 7 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KihS42425hlekYcAzvK:b-xigbys)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
This is looking better. My main concern is that RetryableTxnError and InternalRetryableTxnError are confusingly named. Do we need both, and if so can we better articulate the difference between them? (HandledRetryableTxnError makes more sense, although see my comments below). 

---

Reviewed 18 of 18 files at r6.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 31 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihSdclbNifxLQT2sor:-KihSdclbNifxLQT2sos:b-ujmrte) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/internal/client/sender.go#L31)):*
> ```Go
> }
> 
> // !!! comment
> ```

Don't forget this. 

---

*[pkg/internal/client/txn.go, line 941 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihZFkcDJQ5PPHUP3wI:-KihZFkcDJQ5PPHUP3wJ:bxdqehe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/internal/client/txn.go#L941)):*
> ```Go
> // old epochs and/or the current epoch?
> func (txn *Txn) UpdateStateOnDetachedErr(ctx context.Context, err error, pri roachpb.UserPriority) {
> 	// !!!
> ```

What's the plan here?

---

*[pkg/kv/txn_coord_sender.go, line 516 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kih_lhNwbzXDXhYRKxP:-Kih_lhOLzJR0JZSitdI:bkivwcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L516)):*
> ```Go
> 		abortedErr := roachpb.NewErrorWithTxn(roachpb.NewTransactionAbortedError(), &txn)
> 		retryErr := abortedErr.GoError().(*roachpb.InternalRetryableTxnError)
> 		// TODO(andrei): figure out a UserPriority to use here.
> ```

I think it should be `txnMeta.txn.Priority`. That will be updated when we learn about the current transaction from the server. 

---

*[pkg/kv/txn_coord_sender.go, line 519 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kiha0hBlImTNpzhT8av:-Kiha0hCId2coB6j3-A3:b-n5v4u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L519)):*
> ```Go
> 		newTxn := roachpb.PrepareTransactionForRetry(
> 			retryErr, roachpb.NormalUserPriority, txn.Isolation, txn.Name)
> 		return roachpb.NewError(roachpb.NewHandledRetryableError(
> ```

This process seems awfully convoluted, but I guess it's OK and nicely localized for now.

---

*[pkg/kv/txn_coord_sender.go, line 841 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihbDqecnRIoiIbvD58:-KihbDqecnRIoiIbvD59:b-tibplg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L841)):*
> ```Go
> 		var retryErr *roachpb.InternalRetryableTxnError
> 		if pErr.TransactionRestart != roachpb.TransactionRestart_NONE {
> 			goErr := pErr.GoError()
> ```

It makes me really nervous that this is done in terms of GoError instead of ErrorDetail. 

---

*[pkg/roachpb/data.go, line 973 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXF2JOjjE8HMEA-WA:-KihXF2JOjjE8HMEA-WB:b-kwbh1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/data.go#L973)):*
> ```Go
> // isolation and name help initialize this new transaction.
> //
> // TODO(andrei): This param is confusing: it's passed as the priority of the
> ```

Which param? `pri`?

---

*[pkg/roachpb/errors.go, line 169 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWHt-2rht1A5GgotZ:-KihWHt-2rht1A5Ggot_:b-hqarm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L169)):*
> ```Go
> 	if e.TransactionRestart != TransactionRestart_NONE {
> 		if tErr, ok := e.GetDetail().(*HandledRetryableError); ok {
> 			return &RetryableTxnError{Msg: tErr.Msg, TxnID: tErr.TxnID}
> ```

Why are HandledRetryableErrors turned into RetryableTxnError? HandledRetryableError implements the `error` interface so it looks like it should be fine to return the detail directly here. 

This use of RetryableTxnError seems very different from the use case documented in the comment for NewRetryableTxnError, which seems like it's just for the sql `force_retry` function and other testing.

---

*[pkg/roachpb/errors.go, line 444 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihUbDWAfIAigzlIsjs:-KihUbDWAfIAigzlIsjt:b5c3kjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L444)):*
> ```Go
> func (*HandledRetryableError) canRestartTransaction() TransactionRestart {
> 	// TODO(andrei): make this dependent on the type of the original error
> 	return TransactionRestart_IMMEDIATE
> ```

TransactionRestart_BACKOFF is probably a safer default. Or maybe HandledRetryableError should not implement canRestartTransaction, because the restart has already been "handled" and upper layers don't need to handle it again. 

---

*[pkg/roachpb/errors.proto, line 350 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaJPE5AYmTaJJ3ngqX:-KihWhPnj5Mal0R3B5-d:bu011nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/roachpb/errors.proto#L350)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Could we move towards this design today by having this error contain a `Transaction` and a `bool aborted` instead of either a `Transaction` or a set of `txn_id, retry_priority, retry_timestamp`?  If `aborted` is true, only those fields would be filled in (and `PrepareTransactionForRetry` would know this)
</blockquote></details>

So it's not feasible to use the Transaction here for everything instead of new dedicated fields? 

---

*[pkg/roachpb/errors.proto, line 314 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWqLJCsqPEEeRZNcZ:-KihWqLJCsqPEEeRZNc_:b-e39pdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L314)):*
> ```protobuf
> }
> 
> // InternalRetryableTxnError tells the recipient that the transaction it
> ```

This "Internal" error seems to be used in a lot of places. Maybe we should swap the names of InternalRetryableTxnError and RetryableTxnError? (which at least within the scope of this diff is not used in nearly as many places)

---

*[pkg/roachpb/errors.proto, line 369 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXX5Aluwbi8PdYrWG:-KihXX5Aluwbi8PdYrWH:bktw8ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L369)):*
> ```protobuf
>   optional int32 retry_priority = 4;
>   // If set, a retry should use this timestamp.
>   optional util.hlc.Timestamp retry_timestamp = 5;
> ```

This field doesn't appear to ever be used.

---

*[pkg/sql/distsqlrun/base.go, line 630 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihdexUeeleNZlEKpYK:-KihdexUeeleNZlEKpYL:bvfj4bt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/sql/distsqlrun/base.go#L630)):*
> ```Go
> 		// local TxnCoordSender which always converts these to
> 		// HandledRetryableError. But that's not what we want; we want the global
> 		// TxnCoordSender to do the txn restarting. So I think DistSQL should bypass
> ```

Yeah, that sounds right (assuming "global" means "on the original gateway" and "local" means "local to the distsql worker node"). There should only be one TxnCoordSender involved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-Kihe-ivkLNC2rT-9Lr0:b-wrxiw6)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I got rid of RetryableTxnError. It was technically useful in separating some layers, but I agree that it was confusing. See if you like it better please.

I'm working on the DistSQL code path now... Seeing how to bypass the `TxnCoordSender` on all the nodes but the gateway.

---

Review status: 13 of 32 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/txn.go, line 936 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaMA4K4l6ETtzgJ3Nh:-KikjdXmZJuasQMfl5s-:b-eoai4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/internal/client/txn.go#L936)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It seems like this needs to be resolved before this can be merged. I think we need to make sure that the result of PrepapreTransactionForRetry moves the txn forward in time before assigning it to txn.mu.Proto: it either changes the ID or increases the epoch. 
</blockquote></details>

PrepareTxnForRetry does do what you say. And we now check that the epoch is not stale before updating the state.

---

*[pkg/internal/client/txn_test.go, line 588 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaIQftYORnn4MM-JZ1:-KikgymHKA8ppsYN1lpw:b9zm3z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/internal/client/txn_test.go#L588)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why did this need to be added here? Have we lost something on the non-test path?
</blockquote></details>

this was added because now a call to GoError() does what TxnCoordSender used to do and checks for that observed timestamp. The tests in this file without the TxnCS are pretty hacky, since the client.Txn does rely on the TxnCS to do stuff on the error path. insert rant about cient.Txn and TCS being one thing.

---

*[pkg/kv/txn_coord_sender.go, line 516 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kih_lhNwbzXDXhYRKxP:-Kikhq5rHmsSzpHtqerw:b-yl8egl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L516)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it should be `txnMeta.txn.Priority`. That will be updated when we learn about the current transaction from the server. 
</blockquote></details>

`txnMeta.txn.Priority` is an int; this wants a `UserPriority` which is a class based on which those ints get generated. I don't currently understand this flow of priorities very well...

---

*[pkg/kv/txn_coord_sender.go, line 519 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kiha0hBlImTNpzhT8av:-Kiki4NpLTPudoBWJqME:b-nn3vl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L519)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This process seems awfully convoluted, but I guess it's OK and nicely localized for now.
</blockquote></details>

the real problem, I think, is that the TxnCS and the client.Txn are forced to communicate through the narrow Sender interface.

---

*[pkg/kv/txn_coord_sender.go, line 841 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihbDqecnRIoiIbvD58:-KikSYb3nP7Xt5Ut95Na:brjtk6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L841)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It makes me really nervous that this is done in terms of GoError instead of ErrorDetail. 
</blockquote></details>

the idea is that this code (or at least the
```
newTxn = roachpb.PrepareTransactionForRetry(retryErr, ba.UserPriority, ba.Txn.Isolation, ba.Txn.Name)
```
and the actual update of TxnCoordSender's map will be shared with the DistSQL backdoor path. And DistSQL never has access to a pErr, only to a `pErr.GoError()`. We could perhaps give DistSQL access to the ErrorDetails (if we make `GoError` not convert them to an `InternalRetryableTxnError`), but DistSQL also needs to get its hands on the `pErr.Txn`, and I don't know how I'd do that...

Why does this make you so nervous?

---

*[pkg/roachpb/data.go, line 973 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXF2JOjjE8HMEA-WA:-KikVbnU2Ng0svG2Q1F0:b-ltqe3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/data.go#L973)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Which param? `pri`?
</blockquote></details>

yes, clarified.

---

*[pkg/roachpb/errors.go, line 169 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWHt-2rht1A5GgotZ:-KikVg6CcRSCEmxHIjZ5:bb6cjzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L169)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why are HandledRetryableErrors turned into RetryableTxnError? HandledRetryableError implements the `error` interface so it looks like it should be fine to return the detail directly here. 

This use of RetryableTxnError seems very different from the use case documented in the comment for NewRetryableTxnError, which seems like it's just for the sql `force_retry` function and other testing.
</blockquote></details>

Note that the comment about `NewRetryableTxnError`referred strictly to that constructor (and might have been obsolete by now). `RetryableTxnError` was always used broadly above KV, in SQL.

But I've gotten rid of `RetryableTxnError`.

---

*[pkg/roachpb/errors.go, line 444 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihUbDWAfIAigzlIsjs:-KikW54_Lkkkr6bPzJqk:b-oj8ok6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L444)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

TransactionRestart_BACKOFF is probably a safer default. Or maybe HandledRetryableError should not implement canRestartTransaction, because the restart has already been "handled" and upper layers don't need to handle it again. 
</blockquote></details>

Well the restart has been "handled" in the sense that a `Transaction` proto has been prepared. But higher layers do need to act on it too - the `client.Txn` needs to recognize it and change state, and, in cases when the retry is not handled at the `txn.Exec` level, SQL needs to recognize it and potentially pass it to the SQL client.

But I've removed the `canRestartTransaction` interface from HandledTxnError. It wasn't really needed; layers above TxnCoordSender now only look explicitly for this error, and not other `canRestart` details.

---

*[pkg/roachpb/errors.proto, line 350 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaJPE5AYmTaJJ3ngqX:-Kijwxd1xi5TmYpIeG09:b-54z84p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/roachpb/errors.proto#L350)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

So it's not feasible to use the Transaction here for everything instead of new dedicated fields? 
</blockquote></details>

My TODO was about the communication between storage and KV (through pErr.Transaction), not about this error that's used above storage (so that `pErr.GoError()` wouldn't have to do the error munging that it's doing). So changing anything here wouldn't make me any happy wrt what I was thinking.

As to whether we could use the `Transaction` field here like you're suggesting - I guess we could... But I like it more this way; more explicit... I don't think a partially initialized field, that a client still has operate on, would make anything more easy or clear.
Would you be happier if I turned the 2 alternatives into a `one_of` set?

---

*[pkg/roachpb/errors.proto, line 314 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWqLJCsqPEEeRZNcZ:-KikZism3G1DDukimPU4:b-jt5cu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L314)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This "Internal" error seems to be used in a lot of places. Maybe we should swap the names of InternalRetryableTxnError and RetryableTxnError? (which at least within the scope of this diff is not used in nearly as many places)
</blockquote></details>

got rid of `RetryableTxnError`. I'd keep this one with `Internal`...

---

*[pkg/roachpb/errors.proto, line 369 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXX5Aluwbi8PdYrWG:-Kika3_93enTc_fJTcHQ:bnj904q) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L369)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This field doesn't appear to ever be used.
</blockquote></details>

hmmm yes good catch. I've put this here because, before this change, the `TxnCoordSender` had this line before:
https://github.com/cockroachdb/cockroach/blob/7eb8b75/pkg/kv/txn_coord_sender.go#L853

That line seems nonsensical to me, because a) `pErr.GetTxn()` had already been used in relation to `newTxn` and, b) more importantly, `client.Txn` was ignoring everything but the priority from `pErr` when creating a new `Transaction`. The timestamp was left to be initialized, I guess, on the next `client.Txn.Send()`.

So, I've now removed this field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KikmFsu_F7vaYM1DHto:b-h9d3yi)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 19 of 19 files at r7.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 516 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kih_lhNwbzXDXhYRKxP:-KilPmIRxNMKmGt-GSdo:b-86dsmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L516)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`txnMeta.txn.Priority` is an int; this wants a `UserPriority` which is a class based on which those ints get generated. I don't currently understand this flow of priorities very well...
</blockquote></details>

Ah, I see. I think the right thing to do is to remove the UserPriority argument from PrepareTransactionForRetry, always use the minimum priority, then when the `client.Txn` (which has already stored its UserPriority) gets a handled retryable error it does `txn.UpgradePriority(roachpb.MakePriority(userPriority))` on the error's roachpb.Transaction before using it. 

---

*[pkg/kv/txn_coord_sender.go, line 519 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kiha0hBlImTNpzhT8av:-KilP3viHSGIg5DsSZFT:b-4tloo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L519)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the real problem, I think, is that the TxnCS and the client.Txn are forced to communicate through the narrow Sender interface.
</blockquote></details>

Yeah, I've been questioning the Sender interface for a long time. 

---

*[pkg/kv/txn_coord_sender.go, line 841 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihbDqecnRIoiIbvD58:-KilQSgWrfxCDLw5NEqh:b-rss22n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L841)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the idea is that this code (or at least the
```
newTxn = roachpb.PrepareTransactionForRetry(retryErr, ba.UserPriority, ba.Txn.Isolation, ba.Txn.Name)
```
and the actual update of TxnCoordSender's map will be shared with the DistSQL backdoor path. And DistSQL never has access to a pErr, only to a `pErr.GoError()`. We could perhaps give DistSQL access to the ErrorDetails (if we make `GoError` not convert them to an `InternalRetryableTxnError`), but DistSQL also needs to get its hands on the `pErr.Txn`, and I don't know how I'd do that...

Why does this make you so nervous?

</blockquote></details>

It makes me nervous because it's contrary to the design for roachpb.Error and for error handling in the kv and storage packages. We never used the non-proto RetryableTransactionError in the kv package before. As I've said before, I think it's unfortunate that distsql has diverged from this model, but given that we should still try to keep the error handling model within the kv package the same as it was. 

To that end, my suggestion is to undo the changes to `roachpb.Error.GoError`, and instead change `roachpb.NewError` (or add new `roachpb.Error` constructors) that can reassemble a roachpb.Error from the pieces that distsql has. Then this code and roachpb.PrepareTransactionForRestart would work with roachpb.Errors and ErrorDetails instead of InternalRetryableTxnError.

---

*[pkg/roachpb/errors.go, line 169 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWHt-2rht1A5GgotZ:-KilFv_jEVG5vCRkFoq7:b97pqpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L169)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Note that the comment about `NewRetryableTxnError`referred strictly to that constructor (and might have been obsolete by now). `RetryableTxnError` was always used broadly above KV, in SQL.

But I've gotten rid of `RetryableTxnError`.
</blockquote></details>

I don't think this needs to be a special case any more (now that HandledRetryableTxnError no longer sets e.TransactionRestart); it can just be handled by the `return e.GetDetail()` at the end of the method. 

---

*[pkg/roachpb/errors.go, line 71 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KilFbZOwZzKJ36xnUUV:-KilFbZOwZzKJ36xnUUW:bxdqehe) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ed249465c4e837f68f27da4147121afb3fe036/pkg/roachpb/errors.go#L71)):*
> ```Go
> 	if intErr, ok := err.(*internalError); ok {
> 		*e = *(*Error)(intErr)
> 		// !!!
> ```

What's the plan here?

---

*[pkg/roachpb/errors.proto, line 350 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KiaJPE5AYmTaJJ3ngqX:-KilAj0axXMpU18Y2LIa:b-be43di) ([raw file](https://github.com/cockroachdb/cockroach/blob/7562c4d727439f6424f637be1939900070bd388e/pkg/roachpb/errors.proto#L350)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

My TODO was about the communication between storage and KV (through pErr.Transaction), not about this error that's used above storage (so that `pErr.GoError()` wouldn't have to do the error munging that it's doing). So changing anything here wouldn't make me any happy wrt what I was thinking.

As to whether we could use the `Transaction` field here like you're suggesting - I guess we could... But I like it more this way; more explicit... I don't think a partially initialized field, that a client still has operate on, would make anything more easy or clear.
Would you be happier if I turned the 2 alternatives into a `one_of` set?
</blockquote></details>

It would be a partially-initialized field today, but it might become more fully-initialized in the future (i.e. we could have TxnAbortedError generate the ID for the next attempt and return a fully-initialized proto in addition to the ID of the aborted txn). Using the Txn proto would make it easier to start propagating more information across aborts (like observed timestamps). But I don't feel strongly about this and sticking with separate fields for now is fine too. 

---

*[pkg/roachpb/errors.proto, line 314 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWqLJCsqPEEeRZNcZ:-Kil9FO3FpGJDoobwq0c:b-uy2iof) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L314)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

got rid of `RetryableTxnError`. I'd keep this one with `Internal`...
</blockquote></details>

But what is the scope that it is internal to? It's higher-level than roachpb.Error and is used in more layers instead of being more or less contained within kv and storage. 

---

*[pkg/roachpb/errors.proto, line 369 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXX5Aluwbi8PdYrWG:-KilDRbwSyQhJ35nlZc_:b-15vdyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L369)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

hmmm yes good catch. I've put this here because, before this change, the `TxnCoordSender` had this line before:
https://github.com/cockroachdb/cockroach/blob/7eb8b75/pkg/kv/txn_coord_sender.go#L853

That line seems nonsensical to me, because a) `pErr.GetTxn()` had already been used in relation to `newTxn` and, b) more importantly, `client.Txn` was ignoring everything but the priority from `pErr` when creating a new `Transaction`. The timestamp was left to be initialized, I guess, on the next `client.Txn.Send()`.

So, I've now removed this field.
</blockquote></details>

The linked line was redundant before, but is it now? The `newTxn.Update(errTxn)` line is gone, and I can't find another place where the same thing happens now. 

I think the fact that client.Txn was ignoring the timestamp in the error's txn was a bug, but a small one (the error's txn has a timestamp that we won't get anywhere else, but it's unlikely to be greater than the one we'd get from the hlc on the next call to Send). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 15 of 36 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/sender.go, line 31 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihSdclbNifxLQT2sor:-KimAG0jVoF5BbYO2t1V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/internal/client/sender.go#L31)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Don't forget this. 
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 941 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihZFkcDJQ5PPHUP3wI:-Kim7ufCDRUxCAc5ucJz:b-2jf96t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/internal/client/txn.go#L941)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What's the plan here?
</blockquote></details>

see now pls

---

*[pkg/kv/txn_coord_sender.go, line 516 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-Kih_lhNwbzXDXhYRKxP:-Kim9-Sp3bC0zkQdd4Di:bcuvss7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L516)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Ah, I see. I think the right thing to do is to remove the UserPriority argument from PrepareTransactionForRetry, always use the minimum priority, then when the `client.Txn` (which has already stored its UserPriority) gets a handled retryable error it does `txn.UpgradePriority(roachpb.MakePriority(userPriority))` on the error's roachpb.Transaction before using it. 
</blockquote></details>

I don't know if we should involve the TxnCoordSender in the restart like this. In any case, the UserPriority is not used in the abort case, which is this case.

---

*[pkg/kv/txn_coord_sender.go, line 841 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihbDqecnRIoiIbvD58:-Kim9sCIeG1H4_LDzGcT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/kv/txn_coord_sender.go#L841)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It makes me nervous because it's contrary to the design for roachpb.Error and for error handling in the kv and storage packages. We never used the non-proto RetryableTransactionError in the kv package before. As I've said before, I think it's unfortunate that distsql has diverged from this model, but given that we should still try to keep the error handling model within the kv package the same as it was. 

To that end, my suggestion is to undo the changes to `roachpb.Error.GoError`, and instead change `roachpb.NewError` (or add new `roachpb.Error` constructors) that can reassemble a roachpb.Error from the pieces that distsql has. Then this code and roachpb.PrepareTransactionForRestart would work with roachpb.Errors and ErrorDetails instead of InternalRetryableTxnError.
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.go, line 169 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWHt-2rht1A5GgotZ:-KilZCgEYf79ttDCctzA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.go#L169)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think this needs to be a special case any more (now that HandledRetryableTxnError no longer sets e.TransactionRestart); it can just be handled by the `return e.GetDetail()` at the end of the method. 
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.go, line 71 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KilFbZOwZzKJ36xnUUV:-Kim9tM_otc2Mol9dP-2:b-2jf96t) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ed249465c4e837f68f27da4147121afb3fe036/pkg/roachpb/errors.go#L71)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What's the plan here?
</blockquote></details>

see now pls

---

*[pkg/roachpb/errors.proto, line 314 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWqLJCsqPEEeRZNcZ:-KimLOtEI0eNd2c7x7D6:b-fnb2op) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L314)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

But what is the scope that it is internal to? It's higher-level than roachpb.Error and is used in more layers instead of being more or less contained within kv and storage. 
</blockquote></details>

renamed to... drums... DistSQLRetryableTxnError. Cause DistSQL is the only one that's supposed to ever see it. Curious if you'll dig.

---

*[pkg/roachpb/errors.proto, line 369 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihXX5Aluwbi8PdYrWG:-Kim6aw0Hlcgk13ShN0h:bengrgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L369)):*
I've put the timestamp from the error in the wiped txn, and added a TODO in `client.Txn` about maybe not ignoring it when we initialize the txn proto on the next Send().

> The linked line was redundant before, but is it now? The newTxn.Update(errTxn) line is gone, and I can't find another place where the same thing happens now.

The line `newTxn.Update(errTxn)` is still there - TnxCoordSender.updateState(). the "non-retryable error" cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> (mainly for the kv/client/roachpb parts - you may want someone else to look at the distsql parts)

Thanks for being patient and working through this!

---

Reviewed 21 of 21 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/errors.proto, line 314 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-KihWqLJCsqPEEeRZNcZ:-KimPUNhh3ScT_Y0YDvk:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d89c15cfbb9ee8ab9e8c8afc91d3bf01aed8e61/pkg/roachpb/errors.proto#L314)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

renamed to... drums... DistSQLRetryableTxnError. Cause DistSQL is the only one that's supposed to ever see it. Curious if you'll dig.
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KimWE2DTpCMirFqJNo2:b-8bk81t)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Thanks Ben!

I've added a test on the DistSQL side. Will continue thinking of other testing ideas.

---

Review status: 29 of 38 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15187#-:-KimvEvbYIgOKKDV5QDy:b-umr7i5)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Awesome! Great work!!

vivekmenezes commented:
Great stuff Andrei!

On Fri, Apr 28, 2017, 11:54 AM RaduBerinde <notifications@github.com> wrote:

> Awesome! Great work!!
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/15187#issuecomment-298035864>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBE-lVoOOqYDFmCoUt0sm0G1IsCVbks5r0gujgaJpZM4NCReN>
> .
>
